subsys/pm/device.c

```diff
@@ -401,3 +401,14 @@ int pm_device_driver_init(const struct device *dev,
 	/* Return the PM_DEVICE_ACTION_RESUME result */
 	return rc;
 }
+
+int pm_device_driver_deinit(const struct device *dev,
```

Comment:
This is still confusing to me. If `PM_DEVICE=n`, the function implementation actually does work of de-initialising the device. If `PM_DEVICE=y`, the function is a boolean check of whether it is ready to be de-initialised?

Same function, two very different meanings as far as I can tell.

Reply:
that is correct, the same is the case for pm_device_driver_init():
* If `PM_DEVICE=n`, call on action followed by resume action
* If `PM_DEVICE=y`, init the PM device structure and resume device
* If `PM_DEVICE_RUNTIME=y` and `PM_DEVICE=y` and device has runtime enabled, init the PM device structure but keep device suspended.

We need a common path to deinit/suspend the device if `PM_DEVICE=n`. if `PM_DEVICE=y`, then the user will already have suspended the device with pm_device_action_run() before calling device_deinit(), hence we just need to check that the user actually did this :)

Reply:
Note that `pm_device_driver_deinit()` will only be called as a result of `device_deinit()`, which forces the driver to release all resources, including what was claimed from `ACTION_TURN_ON`. The device will be left in a deinitialized state, meaning `device_is_ready()` returns `false` after `pm_device_driver_deinit()` has been called.

Reply:
> Note that pm_device_driver_deinit() will only be called as a result of device_deinit(), which forces the driver to release all resources, including what was claimed from ACTION_TURN_ON

How is `device_deinit` supposed to know which resources to release? We removed `pm_device_state_get` from drivers for a good reason.

Reply:
A device should not acquire any resources outside of the state `ACTIVE`, and it should release all resources in `SUSPENDED`. The state `OFF` is just there to inform the driver that when this device is turned on again, all hardware registers etc. have been cleared, and you need to reinit them. There is no `ON` state, its just a transition from OFF to suspended, used to initialize the hardware.

Reply:
Check out **Device Behavior for Suspend and Turn Off** section of https://github.com/zephyrproject-rtos/zephyr/issues/84049